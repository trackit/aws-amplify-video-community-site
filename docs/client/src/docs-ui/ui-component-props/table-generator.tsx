import {JsonDocsComponent} from "@stencil/core/internal";
import {h} from "@stencil/core";
import {tableStyle, tableHeaderStyle} from "./ui-component-props.style";
import {
  WebComponentProps,
  TableGenerator,
  DocsBaseEntry,
} from "./ui-component-props.types";

/**
 * Both css and slots only document name and description, so we can share
 * the table generation logic here.
 */
const createBaseTable = <T extends DocsBaseEntry>(entries: T[]) => {
  if (!entries) return undefined;
  const rows = entries.map((entry, index) => {
    const name = entry.name;
    const description = entry.docs;
    if (!name || !description) return undefined;
    return (
      <tr key={index}>
        <td>
          <code>{name}</code>
        </td>
        <td>{description}</td>
      </tr>
    );
  });
  return (
    <div>
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>Description</th>
          </tr>
        </thead>
        <tbody>{rows}</tbody>
      </table>
    </div>
  );
};

/**
 * For attributes, we need to document more information than the base ones,
 * so we create multiple tables.
 */
const attrTableGenerator: TableGenerator = (docs: JsonDocsComponent) => {
  let count = 0;
  const tables = docs?.props.map((prop) => {
    const groupId = `prop-${prop.attr || String(count)}`;
    if (!prop.attr) {
      count++;
    }

    return (
      <div>
        <table class={tableStyle} key={groupId}>
          <thead>
            <tr>
              <th colSpan={2}>
                <div class={tableHeaderStyle}>{prop.name}</div>
              </th>
            </tr>
          </thead>
          <tbody>
            {prop.attr && (
              <tr>
                <th>Attribute</th>
                <td>{prop.attr}</td>
              </tr>
            )}
            <tr>
              <th>Description</th>
              <td>{prop.docs}</td>
            </tr>
            <tr>
              <th>Type</th>
              <td>{prop.type}</td>
            </tr>
            {prop.default && (
              <tr>
                <th>Default</th>
                <td>{prop.default}</td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
    );
  });
  return tables;
};

const cssTableGenerator: TableGenerator = (docs: JsonDocsComponent) => {
  return createBaseTable(docs.styles);
};

const slotsTableGenerator: TableGenerator = (docs: JsonDocsComponent) => {
  return createBaseTable(docs.slots);
};

export const tableGeneratorMap: Record<WebComponentProps, TableGenerator> = {
  attr: attrTableGenerator,
  css: cssTableGenerator,
  slots: slotsTableGenerator,
};
