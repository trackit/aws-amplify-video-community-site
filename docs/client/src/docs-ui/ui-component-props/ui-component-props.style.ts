import {css} from "emotion";

export const tableStyle = css`
  text-align: left;
  width: 100%;
  font-size: 0.875rem;
  margin-bottom: 1rem;

  thead tr {
    background-color: var(--bg-color-tertiary);
  }
  tbody tr th {
    width: 5rem;
  }
`;

export const sectionHeaderStyle = css`
  font-size: 1.2rem;
  font-weight: bold;
  margin-bottom: 1rem;
`;

export const tableHeaderStyle = css`
  font-size: 1rem;
  margin: 0.75rem 0;
`;
