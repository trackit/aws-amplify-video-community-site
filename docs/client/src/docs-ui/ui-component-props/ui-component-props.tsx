import {Component, h, Prop, Host, State} from "@stencil/core";
import docs from "@aws-amplify/ui-components/dist/docs";
import {JsonDocsComponent} from "@stencil/core/internal";
import {tableGeneratorMap} from "./table-generator";
import {WebComponentProps} from "./ui-component-props.types";
import {ATTR_HEADER, CSS_HEADER, SLOTS_HEADER} from "../../constants/strings";

const headerNames: Record<WebComponentProps, string> = {
  attr: ATTR_HEADER,
  css: CSS_HEADER,
  slots: SLOTS_HEADER,
};

@Component({tag: "ui-component-props", shadow: false})
export class DocsUIComponentProps {
  /*** component tag for documented component page */
  @Prop() readonly tag: string;
  /*** whether or not the table contains header tags */
  @Prop() readonly useTableHeaders: boolean = false;
  /** Desired property to document */
  @Prop() readonly propType: WebComponentProps = "attr";

  @State() component: JsonDocsComponent | undefined;

  componentWillRender() {
    this.component = docs.components.find(
      (component) => component.tag === this.tag,
    );
  }

  header() {
    const sectionId = `props-${this.propType}-${this.component?.tag as string}`;
    return this.useTableHeaders ? (
      <docs-in-page-link targetId={sectionId}>
        <h2 id={sectionId}>{headerNames[this.propType]}</h2>
      </docs-in-page-link>
    ) : (
      <h4>{headerNames[this.propType]}</h4>
    );
  }

  content() {
    if (this.propType === "attr") {
      return (
        <p>
          <code>{this.component?.tag}</code>
          &nbsp;provides the following properties to configure the component.
        </p>
      );
    } else if (this.propType === "css") {
      return (
        <p>
          <code>{this.component?.tag}</code>
          &nbsp;provides the following&nbsp;
          <amplify-external-link href="https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties">
            css properties
          </amplify-external-link>
          &nbsp;to modify the style at component level.
        </p>
      );
    } else if (this.propType === "slots") {
      return (
        <p>
          <code>{this.component?.tag}</code>
          &nbsp;provides the following slots based off of the&nbsp;
          <amplify-external-link href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot">
            Web Components slot
          </amplify-external-link>
          &nbsp;element.
        </p>
      );
    }
  }

  propsAreEmpty = () => {
    return (
      (this.propType === "attr" && this.component?.props.length === 0) ||
      (this.propType === "css" && this.component?.styles.length === 0) ||
      (this.propType === "slots" && this.component?.slots.length === 0)
    );
  };

  render() {
    if (!this.component || !this.component.tag || this.propsAreEmpty()) return;

    const tableGenerator = tableGeneratorMap[this.propType];
    return (
      <Host>
        {this.header()}
        {this.content()}
        {tableGenerator(this.component)}
      </Host>
    );
  }
}
