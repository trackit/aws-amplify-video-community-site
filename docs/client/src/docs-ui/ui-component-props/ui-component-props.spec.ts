import {DocsUIComponentProps} from "./ui-component-props";
import {newSpecPage} from "@stencil/core/testing";

describe("ui-component-props", () => {
  describe("Component logic", () => {
    let uiComponentProps: DocsUIComponentProps;
    beforeEach(() => (uiComponentProps = new DocsUIComponentProps()));

    it("should init `tag` as `undefined`", () =>
      expect(uiComponentProps.tag).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsUIComponentProps],
            html: `<ui-component-props />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
