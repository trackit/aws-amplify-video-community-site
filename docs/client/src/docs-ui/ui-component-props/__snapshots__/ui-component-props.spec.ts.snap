// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-component-props Render logic should render 1`] = `<ui-component-props></ui-component-props>`;
